### ISSUE TITLE
[Bug] Loop-carried tile in Vec loses its value across pl.range iterations

### ISSUE THREAD

**Issue by lyfne123** [hw-native-sys/pypto#2481]
## Background

In pypto-lib, software-pipelining the Qwen3-14B decode paged-attention kernel
needs an online-softmax rescale factor produced at stack `k` and applied when
stack `k`'s PV result lands, a few iterations later. Carrying that factor in a
loop-carried tile gives **wrong numerical results**, while the identical
arithmetic without the carry is correct.

Reproduce with:

    python repro_lag_shift.py -p a2a3 -d 0 --init lag0 --steps 4   # PASS  (control)
    python repro_lag_shift.py -p a2a3 -d 0 --init lag1 --steps 4   # FAIL

The repro case is **not committed** — save the file below and run it from
`models/qwen3_14b/` (it imports the repo's `golden` harness).

<details>
<summary>Reproduction case — <code>repro_lag_shift.py</code></summary>

```python
# A loop-carried tile is wrong; the same arithmetic without the carry is right.
#
# All four kernels below compute the SAME recurrence against the SAME golden:
#
#     acc_k = acc_{k-1} * v_k + 1,   v_k = row_max(block k)
#
# They differ only in how v_k reaches the accumulate:
#
#   --init lag0       consumed in the same iteration (no loop-carried tile)  -> PASS
#   --init lag1       carried ONE iteration in `lag1`                        -> FAIL
#   --init distinct   carried TWO iterations in `lag1`/`lag2`                -> FAIL
#   --init aliased    same as distinct, but the two carries start as one
#                     tile -> rejected at codegen with
#                     "tile.move requires distinct source and destination
#                      addresses in Vec, but both resolve to byte offset N"
#
# `lag0` passing is the control: the scaffolding, the golden, `pl.split_aiv`,
# `pl.row_max` and the store are all correct. Introducing the carry is the only
# change, and it makes the result wrong -- already at depth 1.
#
# Observed on pypto origin/main 8e8a3488. Older revisions not checked.
#
#   python repro_lag_shift.py -p a2a3 -d 0 --init lag0 --steps 4   # passes
#   python repro_lag_shift.py -p a2a3 -d 0 --init lag1 --steps 4   # fails
#
# This is the shape a software-pipelined attention kernel needs: an online-softmax
# rescale factor produced at step k and applied when step k's PV result lands,
# LAG iterations later.

import argparse

import torch

import pypto.language as pl
from golden import ScalarSpec, TensorSpec, run_jit

ROWS = 8
COLS = 512
MAX_STEPS = 8
LAG = 2


@pl.jit
def lag_aliased(
    x: pl.Tensor[[MAX_STEPS * ROWS, COLS], pl.FP32],
    out: pl.Out[pl.Tensor[[ROWS, 1], pl.FP32]],
    steps: pl.Scalar[pl.INT32],
) -> pl.Tensor[[ROWS, 1], pl.FP32]:
    with pl.spmd(1, name_hint="lag_aliased"):
        _core = pl.tile.get_block_idx()
        for _aiv in pl.split_aiv(2, mode=pl.SplitMode.NONE):
            zeros = pl.tile.full([ROWS, COLS], dtype=pl.FP32, value=0.0)
            tmp = pl.create_tile([ROWS, COLS], dtype=pl.FP32, target_memory=pl.MemorySpace.Vec)
            # Separate scratch: zrow would otherwise alias `tmp`, which every
            # row_max in the loop overwrites.
            init_tmp = pl.create_tile([ROWS, COLS], dtype=pl.FP32, target_memory=pl.MemorySpace.Vec)
            zrow = pl.tile.muls(pl.row_max(zeros, init_tmp), 0.0)
            acc = pl.tile.adds(zrow, 0.0)
            one = pl.tile.adds(zrow, 1.0)
            lag1 = one
            lag2 = one
            for i in pl.range(steps + LAG):
                v = one
                if i < steps:
                    blk = pl.load(x, [i * ROWS, 0], [ROWS, COLS],
                                  target_memory=pl.MemorySpace.Vec)
                    v = pl.row_max(blk, tmp)
                if i >= LAG:
                    acc = pl.add(pl.mul(acc, lag2), one)
                lag2 = lag1
                lag1 = v
            pl.store(acc, [0, 0], out)
    return out


@pl.jit
def lag_distinct(
    x: pl.Tensor[[MAX_STEPS * ROWS, COLS], pl.FP32],
    out: pl.Out[pl.Tensor[[ROWS, 1], pl.FP32]],
    steps: pl.Scalar[pl.INT32],
) -> pl.Tensor[[ROWS, 1], pl.FP32]:
    with pl.spmd(1, name_hint="lag_distinct"):
        _core = pl.tile.get_block_idx()
        for _aiv in pl.split_aiv(2, mode=pl.SplitMode.NONE):
            zeros = pl.tile.full([ROWS, COLS], dtype=pl.FP32, value=0.0)
            tmp = pl.create_tile([ROWS, COLS], dtype=pl.FP32, target_memory=pl.MemorySpace.Vec)
            # Separate scratch: zrow would otherwise alias `tmp`, which every
            # row_max in the loop overwrites.
            init_tmp = pl.create_tile([ROWS, COLS], dtype=pl.FP32, target_memory=pl.MemorySpace.Vec)
            zrow = pl.tile.muls(pl.row_max(zeros, init_tmp), 0.0)
            acc = pl.tile.adds(zrow, 0.0)
            one = pl.tile.adds(zrow, 1.0)
            lag1 = pl.tile.adds(zrow, 1.0)
            lag2 = pl.tile.adds(zrow, 1.0)
            for i in pl.range(steps + LAG):
                v = one
                if i < steps:
                    blk = pl.load(x, [i * ROWS, 0], [ROWS, COLS],
                                  target_memory=pl.MemorySpace.Vec)
                    v = pl.row_max(blk, tmp)
                if i >= LAG:
                    acc = pl.add(pl.mul(acc, lag2), one)
                lag2 = lag1
                lag1 = v
            pl.store(acc, [0, 0], out)
    return out


@pl.jit
def lag1_distinct(
    x: pl.Tensor[[MAX_STEPS * ROWS, COLS], pl.FP32],
    out: pl.Out[pl.Tensor[[ROWS, 1], pl.FP32]],
    steps: pl.Scalar[pl.INT32],
) -> pl.Tensor[[ROWS, 1], pl.FP32]:
    """Control: identical, but a ONE-deep carry (LAG=1). Same golden."""
    with pl.spmd(1, name_hint="lag1_distinct"):
        _core = pl.tile.get_block_idx()
        for _aiv in pl.split_aiv(2, mode=pl.SplitMode.NONE):
            zeros = pl.tile.full([ROWS, COLS], dtype=pl.FP32, value=0.0)
            tmp = pl.create_tile([ROWS, COLS], dtype=pl.FP32, target_memory=pl.MemorySpace.Vec)
            # Separate scratch: zrow would otherwise alias `tmp`, which every
            # row_max in the loop overwrites.
            init_tmp = pl.create_tile([ROWS, COLS], dtype=pl.FP32, target_memory=pl.MemorySpace.Vec)
            zrow = pl.tile.muls(pl.row_max(zeros, init_tmp), 0.0)
            acc = pl.tile.adds(zrow, 0.0)
            one = pl.tile.adds(zrow, 1.0)
            lag1 = pl.tile.adds(zrow, 1.0)
            for i in pl.range(steps + 1):
                v = one
                if i < steps:
                    blk = pl.load(x, [i * ROWS, 0], [ROWS, COLS],
                                  target_memory=pl.MemorySpace.Vec)
                    v = pl.row_max(blk, tmp)
                if i >= 1:
                    acc = pl.add(pl.mul(acc, lag1), one)
                lag1 = v
            pl.store(acc, [0, 0], out)
    return out


@pl.jit
def lag0_direct(
    x: pl.Tensor[[MAX_STEPS * ROWS, COLS], pl.FP32],
    out: pl.Out[pl.Tensor[[ROWS, 1], pl.FP32]],
    steps: pl.Scalar[pl.INT32],
) -> pl.Tensor[[ROWS, 1], pl.FP32]:
    """Scaffolding control: NO carry at all, consume v in the same iteration."""
    with pl.spmd(1, name_hint="lag0_direct"):
        _core = pl.tile.get_block_idx()
        for _aiv in pl.split_aiv(2, mode=pl.SplitMode.NONE):
            zeros = pl.tile.full([ROWS, COLS], dtype=pl.FP32, value=0.0)
            tmp = pl.create_tile([ROWS, COLS], dtype=pl.FP32, target_memory=pl.MemorySpace.Vec)
            init_tmp = pl.create_tile([ROWS, COLS], dtype=pl.FP32, target_memory=pl.MemorySpace.Vec)
            zrow = pl.tile.muls(pl.row_max(zeros, init_tmp), 0.0)
            acc = pl.tile.adds(zrow, 0.0)
            one = pl.tile.adds(zrow, 1.0)
            for i in pl.range(steps):
                blk = pl.load(x, [i * ROWS, 0], [ROWS, COLS],
                              target_memory=pl.MemorySpace.Vec)
                v = pl.row_max(blk, tmp)
                acc = pl.add(pl.mul(acc, v), one)
            pl.store(acc, [0, 0], out)
    return out


KERNELS = {"lag0": lag0_direct, "aliased": lag_aliased, "distinct": lag_distinct, "lag1": lag1_distinct}


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--platform", default="a2a3")
    parser.add_argument("-d", "--device", type=int, default=0)
    parser.add_argument("--init", choices=list(KERNELS), default="distinct")
    parser.add_argument("--steps", type=int, default=4)
    args = parser.parse_args()

    def golden(values):
        x = values["x"].float()
        acc = torch.zeros([ROWS], dtype=torch.float32)
        for k in range(args.steps):
            v = x[k * ROWS : (k + 1) * ROWS].max(dim=1).values
            acc = acc * v + 1.0
        values["out"][:] = acc.reshape(ROWS, 1)

    torch.manual_seed(0)
    specs = [
        TensorSpec("x", [MAX_STEPS * ROWS, COLS], torch.float32,
                   init_value=lambda: torch.rand([MAX_STEPS * ROWS, COLS]) * 0.5 + 0.5),
        TensorSpec("out", [ROWS, 1], torch.float32, is_output=True),
        ScalarSpec("steps", torch.int32, args.steps),
    ]
    print(f"init={args.init}  steps={args.steps}")
    result = run_jit(
        fn=KERNELS[args.init],
        specs=specs,
        golden_fn=golden,
        runtime_cfg={"platform": args.platform, "device_id": args.device},
        compile_only=args.platform.endswith("sim"),
        rtol=1e-4,
        atol=1e-4,
        save_data=False,
    )
    raise SystemExit(0 if result.passed else 1)
```

</details>

Reproduction environment:

| Component | Version |
|---|---|
| pypto-lib | `48de053` (branch: `fix/replace-compile-for-test-with-lower`) |
| pypto | `8e8a3488` (origin/main) |
| simpler | `93adc386` (detached) |
| ptoas | `0.57` |
| pto-isa | `cd4a3d3f` |
| CANN | `9.0.0` |

All toolchain components match their pins (simpler gitlink, `pto_isa.pin`,
`toolchain/versions.env`).

Diagnosis: **pypto** — a loop-carried tile in Vec memory is not preserved across
`pl.range` iterations.

### Description

All four kernels in the repro compute the same recurrence against the same
golden:

```
acc_k = acc_{k-1} * v_k + 1,   v_k = row_max(block k)
```

They differ only in how `v_k` reaches the accumulate:

| `--init` | how `v_k` reaches the accumulate | result |
|---|---|---|
| `lag0` | consumed in the same iteration, no carry | **PASS** |
| `lag1` | carried ONE iteration in `lag1` | **FAIL** |
| `distinct` | carried TWO iterations in `lag1`/`lag2` | **FAIL** |
| `aliased` | same as `distinct`, carries initialised from one tile | does not compile (below) |

**`lag0` passing is the control**: the scaffolding, the golden, `pl.split_aiv`,
`pl.row_max` and the store are all correct. Introducing the loop-carried tile is
the only change, and it makes the answer wrong — already at depth 1, with no
ordering subtlety to appeal to (`lag1 = v` is a single rebind).

The `aliased` spelling is worth calling out separately because it is the one a
kernel author writes first, and it is rejected at codegen:

```
tile.move requires distinct source and destination addresses in Vec,
but both resolve to byte offset 32800
```

So of the two spellings of a two-deep carry, one does not compile and the other
computes the wrong answer.

Impact: this blocks the PRE_LAUNCH-style skew that a paged-attention kernel needs
to overlap its cube and vector phases; on the real kernel the skew is worth
14-16%.

### Steps to Reproduce

1. Save the file above under `models/qwen3_14b/`.
2. `python repro_lag_shift.py -p a2a3 -d 0 --init lag0 --steps 4` -> PASS.
3. `python repro_lag_shift.py -p a2a3 -d 0 --init lag1 --steps 4` -> FAIL on `out`.
4. `--init aliased` -> `PartialCodegenError` with the `tile.move` message above.

### Expected Behavior

A tile assigned in one iteration and read in a later one keeps its value, so
`lag1` and `distinct` produce the same result as `lag0`.

### Host Platform

`Linux (aarch64)`

### Additional Context

Older pypto revisions were not checked, so I cannot say whether this is a
regression or long-standing.

Related: pypto#2392 (orchestration codegen cross-wires loop-carried `pl.Out`
tensor handles) — same family, different layer: that one is about tensor handles
in orchestration, this one about tiles inside a kernel. pypto#1956 (first-class
`alloc_tile` IR op) touches the same tile-handle machinery.

**Comment by lyfne123** [comment]
Closing: the two defects behind the wrong answers are fixed and verified on a2a3.

The repro turned out to be four independent problems rather than one, which is
why `lag1` and `distinct` failed for different reasons.

**Fixed — `--init lag1` (#2484, 35f4716d).** `YieldFixupMutator` matched an
`IfStmt` branch's yield value with `As<Var>`, which never matches an `IterArg`
(it carries its own `ObjectKind`). The arm that yields the loop's carry
unchanged — the shape `ConvertToSSA` produces for `if cond: acc = f(acc)` —
therefore got no reconciling `tile.move`, while the phi's `return_var` was still
patched onto the sibling arm's buffer. The phi buffer stayed unwritten whenever
that arm ran, and the loop-carry writeback then copied it onto the accumulator.

**Fixed — `--init distinct` (#2494, 4b01d98c).** `pl.yield_` rebinds every carry
simultaneously, but MemoryReuse realized that with `tile.move`s emitted in
iter_arg index order and never checked against each other, so a copy could read
a byte range a preceding copy had just written. For `lag2 = lag1; lag1 = v`,
writing lag1's buffer first made lag2 read the *new* lag1 and the loop carried
`lag2 == lag1`. The copies are now ordered against each other, swaps are broken
with a scratch buffer, and addresses are compared by base+offset rather than by
allocation so slots of one `pl.MemRef(slots=N)` are no longer confused.

**By design — `--init aliased`.** `lag1 = one; lag2 = one` is a name binding, not
an allocation: both carries are the same SSA value, so they get one buffer, and
no ordering can make one iteration preserve two different values in it. This is
now rejected at the loop with an actionable message instead of surfacing
downstream as `tile.move requires distinct source and destination addresses`.
Build each carry with its own op:

```python
lag1 = pl.tile.adds(zrow, 1.0)
lag2 = pl.tile.adds(zrow, 1.0)
```

**Not addressed — `memory_planner=PTOAS`.** Under that planner every variant
still fails to compile with `'pto.tmov' op expects A2/A3 non-mat tmov to use
matching src/dst shapes`. Worth recording for whoever picks it up: `--init lag0`
fails too, and that variant has no conditional and no multi-level carry, so this
is not carry- or branch-specific — it looks like the `[N, 1]` ↔ `[1, N]` reshape
pair that element-wise ops insert around a column vector being lowered to a
`tmov` when ptoas owns placement. Root cause not confirmed on either side. The
PTOAS path is not a priority right now; please open a separate issue if you hit
it.
